src/common/components/dateTimePicker/DateTimePicker.tsx

```diff
@@ -0,0 +1,303 @@
+    true
+  );
+
+  if (!hasAllDateParts) {
+    return;
+  }
+
+  return new Date(Number(Y), Number(M) - 1, Number(D), Number(H), Number(m));
+}
+
+function validateDate(
+  date: Date | undefined | null,
+  dateToCompare: Date | undefined | null,
+  fnc: (date: Date, dateToCompare: Date) => boolean,
+  message: string
+): string {
+  if (!date || !dateToCompare) {
+    return "";
+  }
+
+  const isError = fnc(date, dateToCompare);
+
+  if (isError) {
+    return message;
+  }
+
+  return "";
+}
+
+function useOnOutsideClick<T extends HTMLElement>(
+  callback: () => void
+): Ref<T | null> {
+  const ref = useRef<null | T>(null);
+
+  useEffect(() => {
+    const listener = (event) => {
+      if (ref.current && !ref.current.contains(event.target)) {
+        callback();
+      }
+    };
+
+    document.addEventListener("mousedown", listener);
+    document.addEventListener("touchstart", listener);
+
+    return () => {
+      document.removeEventListener("mousedown", listener);
+      document.removeEventListener("touchstart", listener);
+    };
+  }, [ref, callback]);
+
+  return ref;
+}
+
+type Props = {
+  id: string;
+  name: string;
+  onChange?: (value: Date) => void;
+  value?: Date;
+  label: string;
+  locale?: Locale;
+  minDate?: Date;
+  maxDate?: Date;
+  minDateErrorMessage?: string;
+  maxDateErrorMessage?: string;
+};
+
+export default function DateTimePicker({
+  id,
+  name,
+  label,
+  onChange,
+  value = null,
+  locale = "fi",
+  minDate,
+  maxDate,
+  minDateErrorMessage,
+  maxDateErrorMessage,
+}: Props) {
+  const dateString = getDate(value);
+  const timeString = getTime(value);
+
+  const { t } = useTranslation("date_time_picker");
+  const [isOpen, setIsOpen] = useState<boolean>(false);
+  const [intermediaryDate, setIntermediaryDate] = useIntermediaryState<
+    string | null
+  >(dateString);
+  const [intermediaryTime, setIntermediaryTime] = useIntermediaryState<
+    string | null
+  >(timeString);
+  const [date, setDate] = useState<Date | null>(value);
+  const buttonRef = useRef(null);
+  const containerRef = useOnOutsideClick<HTMLDivElement>(
+    useCallback(() => setIsOpen(false), [])
+  );
+
+  const handleToggleMenu = () => {
+    setIsOpen((previousValue) => !previousValue);
+  };
+
+  const handleCloseMenuButtonClick = () => {
+    setIsOpen(false);
+  };
+
+  const handleDateChange = (nextDate: string) => {
+    setIntermediaryDate(nextDate);
+    onDatePartChange(nextDate, intermediaryTime);
+  };
+
+  const handleTimeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const nextTime = event.target.value;
+
+    setIntermediaryTime(nextTime);
+    onDatePartChange(intermediaryDate, nextTime);
+  };
+
+  const onDatePartChange = (date: string, time: string) => {
+    const dateObject = getDateFromDateAndTimeString(date, time);
+    const isValidDate = dateObject ? getIsDateValid(dateObject) : false;
+
+    if (isValidDate) {
+      setDate(dateObject);
+    }
+  };
```

Comment:
Would useEffect be handy here? No need to manually call `onDatePartChange` in `handleDateChange` and `handleTimeChange`. Those could be then easily inlined in jsx

```
  useEffect(() => {
    const dateObject = getDateFromDateAndTimeString(intermediaryDate, intermediaryTime);
    const isValidDate = dateObject ? getIsDateValid(dateObject) : false;

    if (isValidDate) {
      setDate(dateObject);
    }
  }, [intermediaryDate, intermediaryTime])
  
  // In components props
    onChange={setIntermediaryDate}
    onChange={(e) => setIntermediaryTime(e.target.value)}
```

Reply:
I totally agree with you. I suggested this because I feel that it would perfectly in this situation. At least it is not syncin its own state but rather calling a callback.

I like this solution because we don't have to remember to call onDatePartChange in many places. 

But I'm ok with your solution.